### ISSUE TITLE
(re-) Implement ethernet face

### ISSUE THREAD

**Issue by pulsejet** [named-data/ndnd#70]
There's an implementation in the repo history at https://github.com/named-data/YaNFD/tree/b7be124b6d578ae83b17db1a3e4b75e8402b1e90

`gopacket` should be replaced with the newer one.

**Comment by yoursunny** [comment]
> `gopacket` should be replaced with the newer one.

What newer one? 
It's already using `github.com/gopacket/gopacket` three maintained fork.

**Comment by pulsejet** [comment]
> > `gopacket` should be replaced with the newer one.
> 
> What newer one? It's already using `github.com/gopacket/gopacket` the maintained fork.

Okay, good. I thought it wasn't.

**Comment by yoursunny** [comment]
With `github.com/gopacket/gopacket`, libpcap and C compiler are no longer necessary if `CGO_ENABLED=0` environment is set.
The README should be updated as such.

**Comment by yoursunny** [comment]
This issue was investigated at 17th NDN hackathon.
Here are some notes.

**gopacket/pcap** library:
  * Upon inspection, the old code relies on `gopacket/pcap`, which calls into libpcap and thus requires Cgo.
  * YaNFD is a pure Go application and cannot use Cgo.
  * Thus, it cannot be adopted directly.

**gopacket/afpacket** library:
  * The Ethernet transport can be implemented with `gopacket/afpacket` library.
  * This will only work on Linux.
  * However, `gopacket/afpacket` package has race conditions when trying to close a face (this was noticed during NDN-DPDK development).
  * Due to this race condition, it is only safe to implement multicast Ethernet transport, which does not need to be closed.
  * It is risky to implement unicast Ethernet transport, which must be closable.

**mdlayher/packet** library:
  * I recalled that NDNgo has switched to `mdlayher/packet` library, which is a thin wrapper around AF\_PACKET socket (thus, Linux only).
  * "The connect(2) operation is not supported on packet sockets".
  * A single `packet.Conn` serves both the multicast Ethernet transport and all unicast Ethernet transports on the same netif.
  * This is substantially different from other YaNFD transports, where each transport instance has more-or-less an independent `net.Conn` instance to be used within.

**Comment by zjkmxy** [comment]
We can add conditional compilation for the environments where Cgo is needed.

**Comment by pulsejet** [comment]
> We can add conditional compilation for the environments where Cgo is needed.

Yes, I've no objection to shipping an _additional_ binary for Linux that has cgo dependency.

But from last Friday discussions, seems like we will `wontfix` this (for now at least).
